runtime/doc/lua.txt

```diff
@@ -733,17 +733,222 @@ VIM.JSON                                                            *lua-json*
 
 The *vim.json* module provides encoding and decoding of Lua objects to and
 from JSON-encoded strings. Supports |vim.NIL| and |vim.empty_dict()|.
+This module is provided by `openresty/lua-cjson`. See
+https://github.com/openresty/lua-cjson for more documentation.
 
-vim.json.encode({obj})                                       *vim.json.encode*
+encode({obj})                                              *vim.json.encode()*
     Encodes (or "packs") Lua object {obj} as JSON in a Lua string.
 
-vim.json.decode({str}[, {opts}])                             *vim.json.decode*
+decode({str}[, {opts}])                                    *vim.json.decode()*
     Decodes (or "unpacks") the JSON-encoded {str} to a Lua object.
 
     {opts} is a table with the key `luanil = { object: bool, array: bool }`
     that controls whether `null` in JSON objects or arrays should be converted
     to Lua `nil` instead of `vim.NIL`.
 
+new()                                                         *vim.json.new()*
+    Creates and returns a new `json` object, which is seperate from the
+    global `vim.json`. Can be used to safely configure the options on your
+    object and ensure safe parsing and encoding.
+
+The following methods change the configuration of the json object. ~
+The current settings are always returned. A particular setting is only ~
+changed when the argument is provided (non-nil). ~
+
+NOTE: These are global changes! Be careful with them or else create a new
+object with |vim.json.new()|
+
+                                     *vim.json.encode_empty_table_as_object()*
+encode_empty_table_as_object([{enable}])
+    Default: `false`
+    Parameters: ~
+      • {enable} (boolean / `"on"` / `"off"`) Whether to include array_mt
+        metatable.
+
+    Return: ~
+      Current value: {enable}
+
+    Change the default behavior when encoding an empty Lua table.
+
+    `true` - empty Lua tables are encoded as empty JSON Objects (`{}`).
+    `false` - empty Lua tables are encoded as empty JSON Arrays (`[]`).
+
+   {enable} is either a boolean or a string (`"on"`, `"off"`).
+
+                                       *vim.json.decode_array_with_array_mt()*
```

Comment:
why is this (and others) exposed at all? It's not used in the codebase anywhere.

Reply:
I'm really not sure, but if it's going to be exposed, it should be documented. I just had an issue (https://github.com/codota/tabnine-nvim/issues/65) where one of these options was used and the lack of documentation forced me to read the C source code to determine the exported functions (and the readme of the the *two* repos).

Reply:
It's not explicitly discussed in the original PR, but I think it was deliberate to only officially expose `encode` and `decode` -- it was not the goal to provide a "full" mpack library as part of our standard library. So the first question is whether using anything else _should_ be supported.

Reply:
these functions are necessary for the use of the vim.json module, otherwise, hoops must be jumped through in order to avoid issues (ie, placing otherwise useless properties on a table.)
Another (better) option would be to allow a configuration object to be passed to encode and decode, and in fact I almost attempted that, but I don't know C.

Reply:
> these functions are necessary for the use of the vim.json module, otherwise, hoops must be jumped through in order to avoid issues (ie, placing otherwise useless properties on a table.)


I use `vim.json` in my plugins just fine without these options. Could you explain your use-case where they are necessary?


This https://github.com/codota/tabnine-nvim/issues/65 issue here to me is actually an indication that we should _remove_ these functions.

Letting users change the behavior of the _global_ `vim.json` is just asking for trouble as any plugin can break the behavior of neovim itself (e.g LSP) or other plugins.

Reply:
as of the removal of the `encode_empty_tables_as_objects` from the tabnine code base, we must now ensure that every table always has a property (see codota/tabnine-nvim#67), perhaps a good option would be to remove these methods and instead allow the `encode` and `decode` methods to take a second argument that allows you to set these behaviors, instead of having a global state. or else, if possible, remove these from the global `vim.json` but maintain their existence on any instance created by the user through `vim.json.new()`?

Reply:
If you want to serialize an empty dict, you can use `vim.empty_dict()`.

```lua
vim.print(vim.json.encode({ foo = vim.empty_dict(), bar = {} }))
```

Prints:

```
{"bar":[],"foo":{}}
```


It would really help to explain the use-cases or problems first before focusing on potential solutions.

Reply:
the use case is encoding of values without knowing their contents. If you are expecting an object, but you're given an empty table, you should be able to ensure that your receiver can understand what you pass to it. *Most* internal neovim processes don't need usage of `vim.json`, however, a major use case of this is LSPs, DAPs, and formatters/linters (do these work directly on the file?), which are supposed to work regardless of the client. If an object is expected by the recipient (which can't always be changed), but an table is empty (especially an issue with dynamic objects/state), then it will be encoded differently, which the recipient might not be capable of handling.
Additionally, several of these methods can greatly affect performance and memory usage while encoding/decoding, and certain options (such as `encode_number_precision`, `decode_invalid_numbers`, and `*_max_depth`) can **greatly** affect output, and in the last case, can even completely eliminate output. Some of these options are impossible to work around without knowing *extremely* well the structure of the data, such as the `max_depth` settings, for which, to circumvent, one would have to split the table at a maximum depth recursively (likely splitting into an array of unknown length), then encode each part and attempt to splice the text back together correctly. This would be an extremely complicated situation, which is easily solvable by setting `encode_max_depth` to a larger value. The reverse would also be true, but likely even harder, as it would require (program) understanding of the json structure to split the text, essentially creating a JSON parser in order to *parse JSON* using the built in `vim.json`. While this is a particularly unusual situation (requiring a depth of >100), it is not impossible, and given a dynamic object or an external
source, might be necessary. Provided the option is already there and present in the code base, I'm much more in favor of exporting it in *some manner* (whether through `vim.json.new()` or else a confit option to `vim.json.{encode,decode}`).

Reply:
> the use case is encoding of values without knowing their contents. If you are expecting an object, but you're given an empty table, you should be able to ensure that your receiver can understand what you pass to it. Most internal neovim processes don't need usage of vim.json, however, a major use case of this is LSPs, DAPs, and formatters/linters (do these work directly on the file?), which are supposed to work regardless of the client. If an object is expected by the recipient (which can't always be changed), but an table is empty (especially an issue with dynamic objects/state), then it will be encoded differently, which the recipient might not be capable of handling.

I think the idea here is that `array_mt` is redundant as nvim uses a metatable for empty dict instead, which `vim.json` is configured to use _by default_.

In addition to mfussenegger's example at encoding, this already happens at decoding:
```
lua vim.print(vim.json.decode('{"bar":[],"foo":{}}'))
```
gives
```
{ bar = {},  foo = vim.empty_dict() }
```

Thus we don't need to decorate empty arrays with array_mt _in addition_.

Reply:
ah, yes. Specifically, `array_mt` (and related) can likely be removed with no consequence. in the case of decoding and re-encoding, if `vim.json` returns `vim.empty_dict()` (which can be re-serialized), then it's useless. I thought (falsely?) @mfussenegger was asking the relevance of *all* of these options.